components/feature/push/src/main/java/mozilla/components/feature/push/AutoPushFeature.kt

```diff
@@ -230,6 +230,31 @@ class AutoPushFeature(
         }
     }
 
+    /**
+     * Checks if a subscription for the [scope] already exists.
+     *
+     * @param scope The subscription identifier which usually represents the website's URI.
+     * @param appServerKey An optional key provided by the application server.
+     * @param block The callback invoked when a subscription for the [scope] is found, otherwise null.
+     */
+    fun getSubscription(
+        scope: String,
+        appServerKey: String? = null,
+        block: (AutoPushSubscription?) -> Unit
```

Comment:
That's a fair point that we should invoke the callback on the caller's thread. I'll have to spend some time to figure out how this would affect our integrations.

When we last tried exposing the async functions, it was a bit more complicated to complete the GeckoResult async call with the kotlin async call. We opted to follow the WebExtension learnings since it had a similar one-to-one mapping of APIs.

Now that we have these nested calls where we need to make async checks before another async call, we should re-visit that.

For that reason, I stuck with the current system and hid the `containsSubscription` call internally.

Reply:
We haven't (thankfully) run into issues with being on the wrong thread since consumers always need to be on a background thread to wait for messages, but that doesn't fix the problem with this design that you rightfully spotted.